core/dom/range.js

```diff
@@ -2843,46 +2843,161 @@ CKEDITOR.dom.range = function( root ) {
-		 * As a workaround it will return an array containing only one rectangle which would start in the top left hand corner of the selection and end in the bottom right hand corner.
+		 * As a workaround it will return an array containing only one rectangle which would
+		 * start in the top left hand corner of the selection and end in the bottom right hand corner.
 		 * Possible cases when returned rectangle doesn't fully cover ranges are presented below:
 		 *
 		 * <img src="https://33333.cdn.cke-cs.com/rc1DFuFpHqcR3Mah6y0e/images/90893fcc6c323c10023e73ebfc1fbaa622b48b29c066f7af_ie-rects.png">
 		 *
 		 * @since 4.10.0
+		 * @param {Boolean} [isAbsolute=false] The function will retrieve an absolute
+		 * rectangle of the element i.e. position relative to the upper-left corner of the topmost viewport.
 		 * @returns {CKEDITOR.dom.rect[]}
 		 */
 		getClientRects: ( function() {
 			if ( this.document.getSelection !== undefined ) {
-				return function() {
+				return function( isAbsolute ) {
 					// We need to create native range so we can call native getClientRects.
-					var range = document.createRange(),
-						rectArray;
+					var range = this.root.getDocument().$.createRange(),
+						rectList;
 
 					range.setStart( this.startContainer.$, this.startOffset );
 					range.setEnd( this.endContainer.$, this.endOffset );
 
-					rectArray = CKEDITOR.tools.array.map( range.getClientRects(), function( item ) {
-						return convertRect( item );
-					} );
+					rectList = range.getClientRects();
 
-					// In some cases ( eg. ranges contain only image ) IE will return empty rectList.
-					if ( !range.collapsed && !rectArray.length ) {
-						rectArray = [ getRect( this.createBookmark() ) ];
+					rectList = fixWidgetsRects( rectList, this );
+
+					if ( !rectList.length ) {
+						rectList = fixEmptyRectList( rectList, range, this );
 					}
 
 					range.detach();
 
-					return rectArray;
+					return CKEDITOR.tools.array.map( rectList, function( item ) {
+						return convertRect( item, isAbsolute, this );
+					}, this );
 				};
 			} else {
-				return function() {
-					return [ getRect( this.createBookmark() ) ];
+				return function( isAbsolute ) {
+					return [ convertRect( getRect( this.createBookmark() ), isAbsolute, this ) ];
 				};
 			}
 
+			// Remove all widget rects except for outermost one.
+			function fixWidgetsRects( rectList, context ) {
+				var iterator = context.createIterator(),
+					rectArray = CKEDITOR.tools.array.map( rectList, function( item ) {
+						return item;
+					} ),
+					element,
+					widgetElements = [],
+					widgetRects,
+					widgetRange,
+					nodeList;
+
+				// Lets iterate over each element inside ranges to find widgets.
+				while ( element = iterator.getNextParagraph() ) {
+					nodeList = element.find( '[data-widget]' );
+					// Store each widget element in array.
+					if ( element.hasAttribute( 'data-widget' ) ) {
```

Comment:
`CKEDITOR.dom.node#getAscendant` with second parameter set to `true` should be able to simplify this logic.

Reply:
`getAscendant` returns ancestor, and I'm iterating over children, not sure how I can use this method here.

Reply:
Setting the second parameter to `true` also fetches the element itself. But now I see that this logic is more complicated than I previously thought as it has to find all widgets inside the widget.